controller/backup_controller.go

```diff
@@ -0,0 +1,396 @@
+		baseController: newBaseController("longhorn-backup", logger),
+
+		namespace:    namespace,
+		controllerID: controllerID,
+
+		ds: ds,
+
+		kubeClient:    kubeClient,
+		eventRecorder: eventBroadcaster.NewRecorder(scheme, v1.EventSource{Component: "longhorn-backup-controller"}),
+
+		bvStoreSynced: backupVolumeInformer.Informer().HasSynced,
+		bStoreSynced:  backupInformer.Informer().HasSynced,
+	}
+
+	backupInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    bc.enqueueBackup,
+		UpdateFunc: func(old, cur interface{}) { bc.enqueueBackup(cur) },
+		DeleteFunc: bc.enqueueBackup,
+	})
+
+	return bc
+}
+
+func (bc *BackupController) enqueueBackup(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", obj, err))
+		return
+	}
+
+	bc.queue.AddRateLimited(key)
+}
+
+func (bc *BackupController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer bc.queue.ShutDown()
+
+	bc.logger.Infof("Start Longhorn Backup Snapshot controller")
+	defer bc.logger.Infof("Shutting down Longhorn Backup Snapshot controller")
+
+	if !cache.WaitForNamedCacheSync(bc.name, stopCh, bc.bvStoreSynced, bc.bStoreSynced) {
+		return
+	}
+	for i := 0; i < workers; i++ {
+		go wait.Until(bc.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (bc *BackupController) worker() {
+	for bc.processNextWorkItem() {
+	}
+}
+
+func (bc *BackupController) processNextWorkItem() bool {
+	key, quit := bc.queue.Get()
+	if quit {
+		return false
+	}
+	defer bc.queue.Done(key)
+	err := bc.syncHandler(key.(string))
+	bc.handleErr(err, key)
+	return true
+}
+
+func (bc *BackupController) handleErr(err error, key interface{}) {
+	if err == nil {
+		bc.queue.Forget(key)
+		return
+	}
+
+	if bc.queue.NumRequeues(key) < maxRetries {
+		bc.logger.WithError(err).Warnf("Error syncing Longhorn backup %v", key)
+		bc.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	bc.logger.WithError(err).Warnf("Dropping Longhorn backup %v out of the queue", key)
+	bc.queue.Forget(key)
+}
+
+func (bc *BackupController) syncHandler(key string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "%v: fail to sync backup %v", bc.name, key)
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	if namespace != bc.namespace {
+		return nil
+	}
+	return bc.reconcile(bc.logger, name)
+}
+
+func (bc *BackupController) backupCreation(log logrus.FieldLogger, backup *longhorn.Backup, pollInterval metav1.Duration) error {
+	volumeName, err := bc.getBackupVolume(log, backup)
+	if err != nil {
+		return err
+	}
+
+	backupTarget, err := bc.ds.GetSettingValueExisted(types.SettingNameBackupTarget)
+	if err != nil {
+		return err
+	}
+
+	credential, err := manager.GetBackupCredentialConfig(bc.ds)
+	if err != nil {
+		return err
+	}
+
+	volumeManager := manager.NewVolumeManager(bc.controllerID, bc.ds)
```

Comment:
Do not use package `manager` in `controller`. IMO, that package is for HTTP API handling.

Reply:
I'd prefer to address this in other PR since there the `engine_controller` and `volume_controller` also use package `manager` in `controller`
https://github.com/longhorn/longhorn-manager/blob/f53ebaf/controller/engine_controller.go#L1024
https://github.com/longhorn/longhorn-manager/blob/f53ebaf/controller/engine_controller.go#L1066-L1071
https://github.com/longhorn/longhorn-manager/blob/f53ebaf/controller/volume_controller.go#L2321